## Fashion MNIST

In this Kaggle competition I have used Decision Tree classifier and Ranfom Forest classifier

In [ ]:
import pandas as pd
import numpy as np

importing train data with pandas





In [ ]:
train_data = pd.read_csv("fashion-mnist_train.csv")

checking first 5 elements

In [ ]:
train_data.head()

label  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
0      2       0       0       0       0       0       0       0       0   
1      9       0       0       0       0       0       0       0       0   
2      6       0       0       0       0       0       0       0       5   
3      0       0       0       0       1       2       0       0       0   
4      3       0       0       0       0       0       0       0       0   

   pixel9  ...  pixel775  pixel776  pixel777  pixel778  pixel779  pixel780  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0        30        43         0   
3       0  ...         3         0         0         0         0         1   
4       0  ...         0         0         0         0         0         0   

   pixel781  pixel782  pixel783  pixel784  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 785 columns]

importing test data with pandas

In [ ]:
test_data = pd.read_csv("fashion-mnist_test.csv")

checking last 5 elements

In [ ]:
test_data.tail()

label  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
9995      0       0       0       0       0       0       0       0       0   
9996      6       0       0       0       0       0       0       0       0   
9997      8       0       0       0       0       0       0       0       0   
9998      8       0       1       3       0       0       0       0       0   
9999      1       0       0       0       0       0       0       0     140   

      pixel9  ...  pixel775  pixel776  pixel777  pixel778  pixel779  pixel780  \
9995       0  ...        32        23        14        20         0         0   
9996       0  ...         0         0         0         2        52        23   
9997       0  ...       175       172       172       182       199       222   
9998       0  ...         0         0         0         0         0         1   
9999     119  ...       111        95        75        44         1         0   

      pixel781  pixel782  pixel783  pixel784  
9995         1         0         0         0  
9996        28         0         0         0  
9997        42         0         1         0  
9998         0         0         0         0  
9999         0         0         0         0  

[5 rows x 785 columns]

Dividing data to X and y using drop()

In [ ]:
X_test = test_data.drop(['label'], axis = 1)
y_test = test_data.label

checking if data has NULL elements

In [ ]:
train_data.isnull().sum().sum()

0

In [ ]:
X = train_data.drop(['label'], axis = 1)
y = train_data.label

importing models

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier

Using Grid Search to find best parameters

In [ ]:
parameters = {'max_depth' : [5,10,19]}

In [ ]:
clf_tree = DecisionTreeClassifier()
grid_search_cv_clf_tree = GridSearchCV(clf_tree, parameters, cv = 5)
grid_search_cv_clf_tree.fit(X, y)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best'),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'max_depth': [5, 10, 19]}, pre_dispatch='2*n_jobs',
       refit=True, return_train_score='warn', scoring=None, verbose=0)

In [ ]:
grid_search_cv_clf_tree.best_params_

{'max_depth': 10}

From this, we can see that best depth from [5,10,19] would be 10. I have also checked other numbers, however 10 was optimal.

In [ ]:
best_clf_tree = grid_search_cv_clf_tree.best_estimator_

In [ ]:
best_clf_tree.score(X_test,y_test)

0.8087

Decision Tree Classifier shows 81% accuracy score 

I tried to use Random Forest because it is a composition of "Trees" and the accuracy should be higher this time.

Firstly, I have used default parameters for Randoom Forest

In [ ]:
clf_rf = RandomForestClassifier()

In [ ]:
clf_rf.fit(X,y)

/home/alisher/miniconda3/lib/python2.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [ ]:
clf_rf.score(X,y)

0.9949

In [ ]:
clf_rf.score(X_test,y_test)

0.8645

This showed us 86.45% accuracy

Then, I have used grid search for better solution

In [ ]:
parameters = {'n_estimators':[30, 60, 90], 
             'max_depth': [9, 15, 22]}

In [ ]:
grid_search_cv_clf_rf = GridSearchCV(clf_rf, parameters, cv = 4)

In [ ]:
grid_search_cv_clf_rf.fit(X, y)

GridSearchCV(cv=4, error_score='raise-deprecating',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_estimators': [30, 60, 90], 'max_depth': [9, 15, 22]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [ ]:
grid_search_cv_clf_rf.best_params_

{'max_depth': 22, 'n_estimators': 90}

In [ ]:
best_clf_rf = grid_search_cv_clf_rf.best_estimator_

In [ ]:
best_clf_rf.score(X,y)

0.9960666666666667

In [ ]:
best_clf_rf.score(X_test,y_test)

0.8824

The final result was 88.14% accuracy, while Decision Tree Classifier shows 81% accuracy score. From observations, we can see that Random Forest Classifier works better than Decision Tree classifier for this problem.